tensorflow-examples/src/main/java/org/tensorflow/model/examples/fashionmnist/FashionMnistDataset.java

```diff
@@ -0,0 +1,147 @@
+        try {
+            ByteNdArray trainingImages = readArchive(TRAINING_IMAGES_ARCHIVE);
+            ByteNdArray trainingLabels = readArchive(TRAINING_LABELS_ARCHIVE);
+            ByteNdArray testImages = readArchive(TEST_IMAGES_ARCHIVE);
+            ByteNdArray testLabels = readArchive(TEST_LABELS_ARCHIVE);
+
+            if (validationSize > 0) {
+                return new FashionMnistDataset(
+                        trainingImages.slice(from(validationSize)),
+                        trainingLabels.slice(from(validationSize)),
+                        trainingImages.slice(to(validationSize)),
+                        trainingLabels.slice(to(validationSize)),
+                        testImages,
+                        testLabels
+                );
+            }
+            return new FashionMnistDataset(trainingImages, trainingLabels, null, null, testImages, testLabels);
+
+        } catch (IOException e) {
+            throw new AssertionError(e);
+        }
+    }
+
+    public Iterable<ImageBatch> trainingBatches(int batchSize) {
+        return () -> new ImageBatchIterator(batchSize, trainingImages, trainingLabels);
+    }
+
+    public Iterable<ImageBatch> validationBatches(int batchSize) {
+        return () -> new ImageBatchIterator(batchSize, validationImages, validationLabels);
+    }
+
+    public Iterable<ImageBatch> testBatches(int batchSize) {
+        return () -> new ImageBatchIterator(batchSize, testImages, testLabels);
+    }
+
+    public ImageBatch testBatch() {
+        return new ImageBatch(testImages, testLabels);
+    }
+
+    public long imageSize() {
+        return imageSize;
+    }
+
+    public long numTrainingExamples() {
+        return trainingLabels.shape().size(0);
+    }
+
+    public long numTestingExamples() {
+        return testLabels.shape().size(0);
+    }
+
+    public long numValidationExamples() {
+        return validationLabels.shape().size(0);
+    }
+
+    private final ByteNdArray trainingImages;
+    private final ByteNdArray trainingLabels;
+    private final ByteNdArray validationImages;
+    private final ByteNdArray validationLabels;
+    private final ByteNdArray testImages;
+    private final ByteNdArray testLabels;
+    private final long imageSize;
+
+    private FashionMnistDataset(
+            ByteNdArray trainingImages,
+            ByteNdArray trainingLabels,
+            ByteNdArray validationImages,
+            ByteNdArray validationLabels,
+            ByteNdArray testImages,
+            ByteNdArray testLabels
+    ) {
+        this.trainingImages = trainingImages;
+        this.trainingLabels = trainingLabels;
+        this.validationImages = validationImages;
+        this.validationLabels = validationLabels;
+        this.testImages = testImages;
+        this.testLabels = testLabels;
+        this.imageSize = trainingImages.get(0).shape().size();
+    }
+
+    private static ByteNdArray readArchive(String archiveName) throws IOException {
```

Comment:
Can we use the mnist loader here?

Reply:
Yes, but I tried don't touch the MNISTDataset code here. The best idea - to refactor (convert path to archives to paramters in constructor, for example). In this way, it will have the common loader parametrized by path